crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -25,26 +78,99 @@ use ruff_db::system::{System, SystemPath, SystemPathBuf};
 /// will be `/opt/homebrew`, and `site-packages` will be at
 /// `/opt/homebrew/lib/python3.X/site-packages`.
 ///
-/// This routine does not verify that `sys_prefix_path` points
-/// to an existing directory on disk; it is assumed that this has already
-/// been checked.
-///
 /// [`sys.prefix`]: https://docs.python.org/3/library/sys.html#sys.prefix
+#[derive(Debug, PartialEq, Eq)]
+struct SysPrefixPath(SystemPathBuf);
+
+impl SysPrefixPath {
+    fn new(unvalidated_path: impl AsRef<SystemPath>, system: &dyn System) -> Option<Self> {
+        let unvalidated_path = unvalidated_path.as_ref();
+        // It's important to resolve symlinks here rather than simply making the path absolute,
+        // since system Python installations often only put symlinks in the "expected"
+        // locations for `home` and `site-packages`
+        let canonicalized = system.canonicalize_path(unvalidated_path).ok()?;
+        system
+            .is_directory(&canonicalized)
+            .then_some(Self(canonicalized))
+    }
```

Comment:
Same as above. 

* I think we should either propagate a canonicalize failure or at least log a warning if it is safe to recover. 
* Can `unvalidated_path` be relative, if so, what to is it relative?